File: services/healthreport/app/collectors/victoria.py
"""Host and container facts from VictoriaMetrics.

Grafana Alloy already ships node, cadvisor and CrowdSec metrics from mljr and
nuc (and nas once services/nas-alloy is deployed), so nearly everything here
is a PromQL query rather than a login to a box.
"""

from ..model import CollectorResult, Observation
from .base import collector, http_get

# Pseudo-filesystems that are always ~100% full or always empty, and say
# nothing about the health of the machine.
FS_EXCLUDE = 'fstype!~"tmpfs|overlay|squashfs|ramfs|devtmpfs|fuse.*|iso9660|autofs"'


def query(config, promql):
    """Run an instant query. Returns a list of (labels, float value)."""
    response = http_get(
        config.victoria_url.rstrip("/") + "/api/v1/query",
        params={"query": promql},
        timeout=30,
    )
    payload = response.json()
    if payload.get("status") != "success":
        raise RuntimeError("query failed: %s" % payload.get("error", "unknown"))
    out = []
    for series in payload.get("data", {}).get("result", []):
        try:
            out.append((series.get("metric", {}), float(series["value"][1])))
        except (KeyError, IndexError, ValueError):
            continue
    return out


def _grafana(config):
    return {"grafana_url": config.grafana_url}


@collector("host_metrics")
def collect(config, rules):
    result = CollectorResult(name="host_metrics")
    obs = result.observations
    data = {}

    # --- filesystems -------------------------------------------------------
    promql = (
        "100 - (node_filesystem_avail_bytes{%s} / node_filesystem_size_bytes{%s} * 100)"
        % (FS_EXCLUDE, FS_EXCLUDE)
    )
    rows = query(config, promql)
    data["filesystems"] = [
        {"instance": m.get("instance"), "mountpoint": m.get("mountpoint"), "used_percent": round(v, 2)}
        for m, v in rows
    ]
    for metric, value in rows:
        host = metric.get("instance", "unknown")
        mount = metric.get("mountpoint", "?")
        obs.append(Observation(
            id="disk_usage.%s.%s" % (host, mount),
            collector="host_metrics",
            subject=host,
            kind="disk_usage",
            value=round(value, 2),
            unit="percent",
            message="%s %s at %.1f%% used" % (host, mount, value),
            evidence=dict(promql=promql, **_grafana(config)),
        ))

    # --- projected exhaustion ---------------------------------------------
    # predict_linear over a week, extrapolated a fortnight ahead. Negative
    # means the filesystem is on track to be full before then.
    promql = (
        "predict_linear(node_filesystem_avail_bytes{%s}[7d], 14 * 86400)" % FS_EXCLUDE
    )
    for metric, value in query(config, promql):
        if value >= 0:
            continue
        host = metric.get("instance", "unknown")
        mount = metric.get("mountpoint", "?")
        obs.append(Observation(
            id="disk_fill_projection.%s.%s" % (host, mount),
            collector="host_metrics",
            subject=host,
            kind="disk_fill_projection",
            value=round(value / (1024.0 ** 3), 2),
            unit="gib_projected_free",
            message="%s %s is projected to run out of space within 14 days" % (host, mount),
            evidence=dict(promql=promql, **_grafana(config)),
        ))

    # --- memory ------------------------------------------------------------
    promql = "100 * (1 - node_memory_MemAvailable_bytes / node_memory_MemTotal_bytes)"
    rows = query(config, promql)
    data["memory"] = [{"instance": m.get("instance"), "used_percent": round(v, 2)} for m, v in rows]
    for metric, value in rows:
        host = metric.get("instance", "unknown")
        obs.append(Observation(
            id="memory_usage.%s." % host,
            collector="host_metrics",
            subject=host,
            kind="memory_usage",
            value=round(value, 2),
            unit="percent",
            message="%s memory at %.1f%% used" % (host, value),
            evidence=dict(promql=promql, **_grafana(config)),
        ))

    # --- load per core -----------------------------------------------------
    promql = (
        "node_load15 / on(instance) group_left "
        'count by(instance) (node_cpu_seconds_total{mode="idle"})'
    )
    rows = query(config, promql)
    data["load_per_core"] = [{"instance": m.get("instance"), "load15_per_core": round(v, 2)} for m, v in rows]
    for metric, value in rows:
        host = metric.get("instance", "unknown")
        obs.append(Observation(
            id="load_per_core.%s." % host,
            collector="host_metrics",
            subject=host,
            kind="load_per_core",
            value=round(value, 2),
            unit="ratio",
            message="%s 15m load is %.2f per core" % (host, value),
            evidence=dict(promql=promql, **_grafana(config)),
        ))

    # --- CPU steal (VPS noisy-neighbour detection) -------------------------
    promql = '100 * avg by(instance) (rate(node_cpu_seconds_total{mode="steal"}[1h]))'
    for metric, value in query(config, promql):
        host = metric.get("instance", "unknown")
        obs.append(Observation(
            id="cpu_steal.%s." % host,
            collector="host_metrics",
            subject=host,
            kind="cpu_steal",
            value=round(value, 2),
            unit="percent",
            message="%s CPU steal at %.1f%% over the last hour" % (host, value),
            evidence=dict(promql=promql, **_grafana(config)),
        ))

    # --- unexpected reboot -------------------------------------------------
    promql = "time() - node_boot_time_seconds"
    rows = query(config, promql)
    data["uptime_seconds"] = [{"instance": m.get("instance"), "uptime": int(v)} for m, v in rows]
    for metric, value in rows:
        host = metric.get("instance", "unknown")
        obs.append(Observation(
            id="recent_reboot.%s." % host,
            collector="host_metrics",
            subject=host,
            kind="recent_reboot",
            value=int(value),
            unit="seconds_since_boot",
            message="%s booted %.1f hours ago" % (host, value / 3600.0),
            evidence=dict(promql=promql, **_grafana(config)),
        ))

    # --- failed systemd units ---------------------------------------------
    # Requires enable_collectors = ["systemd"] in the Alloy config.
    promql = 'node_systemd_unit_state{state="failed"} == 1'
    for metric, _value in query(config, promql):
        host = metric.get("instance", "unknown")
        unit = metric.get("name", "unknown")
        obs.append(Observation(
            id="systemd_failed.%s.%s" % (host, unit),
            collector="host_metrics",
            subject=host,
            kind="systemd_failed",
            value=unit,
            message="%s: systemd unit %s is failed" % (host, unit),
            evidence=dict(promql=promql, **_grafana(config)),
        ))

    # --- metric staleness --------------------------------------------------
    # A host that stopped reporting is invisible to every query above, so it
    # must be checked explicitly against the expected host list.
    promql = "time() - max by(instance) (timestamp(node_load1))"
    seen = {}
    for metric, value in query(config, promql):
        seen[metric.get("instance", "unknown")] = value
    data["metrics_age_seconds"] = {k: round(v, 1) for k, v in seen.items()}
    for host in config.all_hosts():
        age = seen.get(host)
        if age is None:
            obs.append(Observation(
                id="metrics_missing.%s." % host,
                collector="host_metrics",
                subject=host,
                kind="metrics_missing",
                value=None,
                message="%s is not reporting any metrics" % host,
                evidence=dict(promql=promql, **_grafana(config)),
            ))
        else:
            obs.append(Observation(
                id="metrics_stale.%s." % host,
                collector="host_metrics",
                subject=host,
                kind="metrics_stale",
                value=round(age, 1),
                unit="seconds",
                message="%s metrics are %.0f seconds old" % (host, age),
                evidence=dict(promql=promql, **_grafana(config)),
            ))

    result.data = data
    return result


@collector("containers")
def collect_containers(config, rules):
    """Container inventory drift and restart loops.

    The important query is the inventory diff: a container that existed 24h ago
    and does not exist now is exactly the silent failure this report is for.
    """
    result = CollectorResult(name="containers")
    obs = result.observations
    lookback = "%dh" % config.lookback_hours

    now_q = 'count by(instance, name) (container_last_seen{name!=""})'
    then_q = 'count by(instance, name) (container_last_seen{name!=""} offset %s)' % lookback

    current = {(m.get("instance"), m.get("name")) for m, _ in query(config, now_q)}
    previous = {(m.get("instance"), m.get("name")) for m, _ in query(config, then_q)}

    # Only compare hosts present on both sides of the window. This suppresses
    # two false-positive floods: a host that was down for the whole lookback,
    # and the one-off churn after an `instance` label is rewritten (as happened
    # when mljr stopped reporting as vmi2945702.contaboserver.net).
    comparable = {h for h, _ in current} & {h for h, _ in previous}
    current = {(h, n) for h, n in current if h in comparable}
    previous = {(h, n) for h, n in previous if h in comparable}

    for host, name in sorted(previous - current):
        obs.append(Observation(
            id="container_missing.%s.%s" % (host, name),
            collector="containers",
            subject=host or "unknown",
            kind="container_missing",
            value=name,
            message="%s: container %s was running %s ago and is gone now" % (host, name, lookback),
            evidence={"promql": then_q, "grafana_url": config.grafana_url},
        ))

    for host, name in sorted(current - previous):
        obs.append(Observation(
            id="container_new.%s.%s" % (host, name),
            collector="containers",
            subject=host or "unknown",
            kind="container_new",
            value=name,
            message="%s: container %s appeared in the last %s" % (host, name, lookback),
            evidence={"promql": now_q},
        ))

    restart_q = 'changes(container_start_time_seconds{name!=""}[%s])' % lookback
    for metric, value in query(config, restart_q):
        if value <= 1:
            continue
        host = metric.get("instance", "unknown")
        name = metric.get("name", "unknown")
        obs.append(Observation(
            id="container_restarts.%s.%s" % (host, name),
            collector="containers",
            subject=host,
            kind="container_restarts",
            value=int(value),
            unit="restarts",
            message="%s: container %s restarted %d times in %s" % (host, name, value, lookback),
            evidence={"promql": restart_q, "grafana_url": config.grafana_url},
        ))

    oom_q = "increase(container_oom_events_total[%s])" % lookback
    for metric, value in query(config, oom_q):
        if value < 1:
            continue
        host = metric.get("instance", "unknown")
        name = metric.get("name", "unknown")
        obs.append(Observation(
            id="container_oom.%s.%s" % (host, name),
            collector="containers",
            subject=host,
            kind="container_oom",
            value=int(value),
            unit="events",
            message="%s: container %s was OOM-killed %d times in %s" % (host, name, value, lookback),
            evidence={"promql": oom_q, "grafana_url": config.grafana_url},
        ))

    result.data = {
        "current_count": len(current),
        "previous_count": len(previous),
        "missing": sorted("%s/%s" % (h, n) for h, n in previous - current),
        "new": sorted("%s/%s" % (h, n) for h, n in current - previous),
    }
    return result

File: tools/internal/healthreport/collectors/victoria.go
// Host and container facts from VictoriaMetrics. Grafana Alloy already
// ships node, cadvisor and CrowdSec metrics from mljr and nuc (and nas once
// services/nas-alloy is deployed), so nearly everything here is a PromQL
// query rather than a login to a box.
package collectors

import (
	"encoding/json"
	"fmt"
	"net/url"
	"regexp"
	"sort"
	"strconv"
	"time"

	hr "github.com/MrCodeEU/homelab-automation/tools/internal/healthreport"
)

// Session-scope units flap on every SSH/console login and carry no signal.
var systemdNoise = regexp.MustCompile(`^session-c?\d+\.scope$`)

// Pseudo-filesystems that are always ~100% full or always empty, and say
// nothing about the health of the machine.
const fsExclude = `fstype!~"tmpfs|overlay|squashfs|ramfs|devtmpfs|fuse.*|iso9660|autofs"`

type diskFillKey struct{ host, mount string }

// Retired-but-kept mounts: not receiving new data, so predict_linear's
// 7-day window keeps extrapolating a one-time historical drop as an
// ongoing trend forever.
var diskFillProjectionExclude = map[diskFillKey]bool{{"nas", "/mnt/cache"}: true}

type vmSample struct {
	metric map[string]string
	value  float64
}

func vmQuery(cfg hr.Config, promql string) ([]vmSample, error) {
	resp, err := httpGet(
		cfg.VictoriaURL+"/api/v1/query?"+url.Values{"query": {promql}}.Encode(),
		30*time.Second, nil,
	)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	var payload struct {
		Status string `json:"status"`
		Error  string `json:"error"`
		Data   struct {
			Result []struct {
				Metric map[string]string `json:"metric"`
				Value  [2]any            `json:"value"`
			} `json:"result"`
		} `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&payload); err != nil {
		return nil, err
	}
	if payload.Status != "success" {
		return nil, fmt.Errorf("query failed: %s", payload.Error)
	}
	out := make([]vmSample, 0, len(payload.Data.Result))
	for _, series := range payload.Data.Result {
		s, ok := series.Value[1].(string)
		if !ok {
			continue
		}
		v, err := strconv.ParseFloat(s, 64)
		if err != nil {
			continue
		}
		out = append(out, vmSample{metric: series.Metric, value: v})
	}
	return out, nil
}

func metricHost(m map[string]string) string {
	if v, ok := m["instance"]; ok {
		return v
	}
	return "unknown"
}

func round2(v float64) float64 { return float64(int64(v*100+sign(v)*0.5)) / 100 }
func sign(v float64) float64 {
	if v < 0 {
		return -1
	}
	return 1
}

func init() {
	hr.RegisterCollector("host_metrics", collectHostMetrics)
	hr.RegisterCollector("containers", collectContainers)
}

func collectHostMetrics(cfg hr.Config, rules hr.RulesFile) *hr.CollectorResult {
	result := hr.NewCollectorResult("host_metrics")
	data := &hr.HostMetricsData{MetricsAgeSeconds: map[string]float64{}}

	grafanaEvidence := func(promql string) map[string]any {
		return map[string]any{"promql": promql, "grafana_url": cfg.GrafanaURL}
	}

	// --- filesystems ---
	promql := fmt.Sprintf(
		"100 - (node_filesystem_avail_bytes{%s} / node_filesystem_size_bytes{%s} * 100)", fsExclude, fsExclude,
	)
	rows, err := vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := metricHost(s.metric)
		mount := s.metric["mountpoint"]
		data.Filesystems = append(data.Filesystems, hr.FilesystemRow{Instance: host, Mountpoint: mount, UsedPercent: round2(s.value)})
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("disk_usage.%s.%s", host, mount), Collector: "host_metrics",
			Subject: host, Kind: "disk_usage", Value: round2(s.value), Unit: "percent",
			Message: fmt.Sprintf("%s %s at %.1f%% used", host, mount, s.value),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- projected exhaustion ---
	promql = fmt.Sprintf("predict_linear(node_filesystem_avail_bytes{%s}[7d], 14 * 86400)", fsExclude)
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		if s.value >= 0 {
			continue
		}
		host := metricHost(s.metric)
		mount := s.metric["mountpoint"]
		if diskFillProjectionExclude[diskFillKey{host, mount}] {
			continue
		}
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("disk_fill_projection.%s.%s", host, mount), Collector: "host_metrics",
			Subject: host, Kind: "disk_fill_projection", Value: round2(s.value / (1024.0 * 1024.0 * 1024.0)),
			Unit:     "gib_projected_free",
			Message:  fmt.Sprintf("%s %s is projected to run out of space within 14 days", host, mount),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- memory ---
	promql = "100 * (1 - node_memory_MemAvailable_bytes / node_memory_MemTotal_bytes)"
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := metricHost(s.metric)
		data.Memory = append(data.Memory, hr.MemoryRow{Instance: host, UsedPercent: round2(s.value)})
		result.Observations = append(result.Observations, &hr.Observation{
			ID: "memory_usage." + host + ".", Collector: "host_metrics", Subject: host, Kind: "memory_usage",
			Value: round2(s.value), Unit: "percent",
			Message: fmt.Sprintf("%s memory at %.1f%% used", host, s.value),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- load per core ---
	promql = `node_load15 / on(instance) group_left count by(instance) (node_cpu_seconds_total{mode="idle"})`
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := metricHost(s.metric)
		data.LoadPerCore = append(data.LoadPerCore, hr.LoadRow{Instance: host, Load15PerCore: round2(s.value)})
		result.Observations = append(result.Observations, &hr.Observation{
			ID: "load_per_core." + host + ".", Collector: "host_metrics", Subject: host, Kind: "load_per_core",
			Value: round2(s.value), Unit: "ratio",
			Message: fmt.Sprintf("%s 15m load is %.2f per core", host, s.value),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- CPU steal (VPS noisy-neighbour detection) ---
	promql = `100 * avg by(instance) (rate(node_cpu_seconds_total{mode="steal"}[1h]))`
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := metricHost(s.metric)
		result.Observations = append(result.Observations, &hr.Observation{
			ID: "cpu_steal." + host + ".", Collector: "host_metrics", Subject: host, Kind: "cpu_steal",
			Value: round2(s.value), Unit: "percent",
			Message: fmt.Sprintf("%s CPU steal at %.1f%% over the last hour", host, s.value),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- unexpected reboot ---
	promql = "time() - node_boot_time_seconds"
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := metricHost(s.metric)
		data.UptimeSeconds = append(data.UptimeSeconds, hr.UptimeRow{Instance: host, Uptime: int64(s.value)})
		result.Observations = append(result.Observations, &hr.Observation{
			ID: "recent_reboot." + host + ".", Collector: "host_metrics", Subject: host, Kind: "recent_reboot",
			Value: int64(s.value), Unit: "seconds_since_boot",
			Message: fmt.Sprintf("%s booted %.1f hours ago", host, s.value/3600.0),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- failed systemd units ---
	promql = `node_systemd_unit_state{state="failed"} == 1`
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := metricHost(s.metric)
		unit := s.metric["name"]
		if unit == "" {
			unit = "unknown"
		}
		if systemdNoise.MatchString(unit) {
			continue
		}
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("systemd_failed.%s.%s", host, unit), Collector: "host_metrics",
			Subject: host, Kind: "systemd_failed", Value: unit,
			Message:  fmt.Sprintf("%s: systemd unit %s is failed", host, unit),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- metric staleness ---
	promql = "time() - max by(instance) (timestamp(node_load1))"
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	seenAge := map[string]float64{}
	for _, s := range rows {
		seenAge[metricHost(s.metric)] = s.value
	}
	for host, age := range seenAge {
		data.MetricsAgeSeconds[host] = round2(age)
	}
	for _, host := range cfg.AllHosts() {
		age, ok := seenAge[host]
		if !ok {
			result.Observations = append(result.Observations, &hr.Observation{
				ID: "metrics_missing." + host + ".", Collector: "host_metrics", Subject: host, Kind: "metrics_missing",
				Value:    nil,
				Message:  fmt.Sprintf("%s is not reporting any metrics", host),
				Evidence: grafanaEvidence(promql), Severity: "info",
			})
		} else {
			result.Observations = append(result.Observations, &hr.Observation{
				ID: "metrics_stale." + host + ".", Collector: "host_metrics", Subject: host, Kind: "metrics_stale",
				Value: round2(age), Unit: "seconds",
				Message:  fmt.Sprintf("%s metrics are %.0f seconds old", host, age),
				Evidence: grafanaEvidence(promql), Severity: "info",
			})
		}
	}

	// --- SMART health/temperature ---
	promql = "smartctl_device_smart_status"
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := smartHost(s.metric)
		device := s.metric["device"]
		if device == "" {
			device = "unknown"
		}
		passed := s.value != 0
		status := "passed"
		if !passed {
			status = "FAILED"
		}
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("smart_health.%s.%s", host, device), Collector: "host_metrics",
			Subject: host, Kind: "smart_health", Value: passed,
			Message:  fmt.Sprintf("%s %s SMART self-assessment %s", host, device, status),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	promql = `smartctl_device_temperature{temperature_type="current"}`
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := smartHost(s.metric)
		device := s.metric["device"]
		if device == "" {
			device = "unknown"
		}
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("disk_temperature.%s.%s", host, device), Collector: "host_metrics",
			Subject: host, Kind: "disk_temperature", Value: int64(s.value), Unit: "celsius",
			Message:  fmt.Sprintf("%s %s at %d C", host, device, int64(s.value)),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- btrfs pool errors ---
	promql = "node_btrfs_device_errors_total"
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	type poolKey struct{ host, device string }
	poolErrors := map[poolKey]map[string]float64{}
	var poolOrder []poolKey
	for _, s := range rows {
		if s.value <= 0 {
			continue
		}
		host := smartHost(s.metric)
		device := s.metric["device"]
		if device == "" {
			device = "unknown"
		}
		errType := s.metric["type"]
		if errType == "" {
			errType = "unknown"
		}
		key := poolKey{host, device}
		entry, ok := poolErrors[key]
		if !ok {
			entry = map[string]float64{}
			poolErrors[key] = entry
			poolOrder = append(poolOrder, key)
		}
		entry[errType] += s.value
	}
	for _, key := range poolOrder {
		errs := poolErrors[key]
		var types []string
		for t := range errs {
			types = append(types, t)
		}
		sort.Strings(types)
		total := 0.0
		var parts []string
		for _, t := range types {
			total += errs[t]
			parts = append(parts, fmt.Sprintf("%s=%d", t, int64(errs[t])))
		}
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("btrfs_pool_errors.%s.%s", key.host, key.device), Collector: "host_metrics",
			Subject: key.host, Kind: "btrfs_pool_errors", Value: int64(total), Unit: "errors",
			Message:  fmt.Sprintf("%s: btrfs device %s has errors: %s", key.host, key.device, joinComma(parts)),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	// --- CrowdSec (mljr only) ---
	promql = "sum by(host) (cs_active_decisions)"
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		host := smartHost(s.metric)
		result.Observations = append(result.Observations, &hr.Observation{
			ID: "crowdsec_active_decisions." + host + ".", Collector: "host_metrics", Subject: host,
			Kind: "crowdsec_active_decisions", Value: int64(s.value), Unit: "decisions",
			Message:  fmt.Sprintf("%s: %d active CrowdSec ban decision(s)", host, int64(s.value)),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	promql = "sum by(host) (increase(cs_alerts[24h]))"
	rows, err = vmQuery(cfg, promql)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		if s.value <= 0 {
			continue
		}
		host := smartHost(s.metric)
		result.Observations = append(result.Observations, &hr.Observation{
			ID: "crowdsec_alerts_24h." + host + ".", Collector: "host_metrics", Subject: host,
			Kind: "crowdsec_alerts_24h", Value: int64(s.value), Unit: "alerts",
			Message:  fmt.Sprintf("%s: %d CrowdSec alert(s) in the last 24h", host, int64(s.value)),
			Evidence: grafanaEvidence(promql), Severity: "info",
		})
	}

	result.Data = data
	return result
}

func smartHost(m map[string]string) string {
	if v, ok := m["host"]; ok {
		return v
	}
	return metricHost(m)
}

func joinComma(parts []string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += ", "
		}
		out += p
	}
	return out
}

// collectContainers: container inventory drift and restart loops. The
// important query is the inventory diff: a container that existed 24h ago
// and does not exist now is exactly the silent failure this report is for.
func collectContainers(cfg hr.Config, rules hr.RulesFile) *hr.CollectorResult {
	result := hr.NewCollectorResult("containers")
	lookback := fmt.Sprintf("%dh", cfg.LookbackHours)

	nowQ := `count by(instance, name) (container_last_seen{name!=""})`
	thenQ := fmt.Sprintf(`count by(instance, name) (container_last_seen{name!=""} offset %s)`, lookback)

	type key struct{ host, name string }
	inventory := func(promql string) map[key]bool {
		rows, err := vmQuery(cfg, promql)
		if err != nil {
			panic(err)
		}
		out := map[key]bool{}
		for _, s := range rows {
			name := s.metric["name"]
			if hr.IsEphemeral(name) {
				continue
			}
			out[key{s.metric["instance"], name}] = true
		}
		return out
	}

	current := inventory(nowQ)
	previous := inventory(thenQ)

	// Only compare hosts present on both sides of the window.
	currentHosts := map[string]bool{}
	for k := range current {
		currentHosts[k.host] = true
	}
	previousHosts := map[string]bool{}
	for k := range previous {
		previousHosts[k.host] = true
	}
	comparable := map[string]bool{}
	for h := range currentHosts {
		if previousHosts[h] {
			comparable[h] = true
		}
	}
	filterComparable := func(m map[key]bool) map[key]bool {
		out := map[key]bool{}
		for k := range m {
			if comparable[k.host] {
				out[k] = true
			}
		}
		return out
	}
	current = filterComparable(current)
	previous = filterComparable(previous)

	var missing, added []key
	for k := range previous {
		if !current[k] {
			missing = append(missing, k)
		}
	}
	for k := range current {
		if !previous[k] {
			added = append(added, k)
		}
	}
	sort.Slice(missing, func(i, j int) bool { return keyLess(missing[i], missing[j]) })
	sort.Slice(added, func(i, j int) bool { return keyLess(added[i], added[j]) })

	for _, k := range missing {
		host := k.host
		if host == "" {
			host = "unknown"
		}
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("container_missing.%s.%s", k.host, k.name), Collector: "containers",
			Subject: host, Kind: "container_missing", Value: k.name,
			Message:  fmt.Sprintf("%s: container %s was running %s ago and is gone now", k.host, k.name, lookback),
			Evidence: map[string]any{"promql": thenQ, "grafana_url": cfg.GrafanaURL}, Severity: "info",
		})
	}
	for _, k := range added {
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("container_new.%s.%s", k.host, k.name), Collector: "containers",
			Subject: k.host, Kind: "container_new", Value: k.name,
			Message:  fmt.Sprintf("%s: container %s appeared in the last %s", k.host, k.name, lookback),
			Evidence: map[string]any{"promql": nowQ}, Severity: "info",
		})
	}

	restartQ := fmt.Sprintf(`changes(container_start_time_seconds{name!=""}[%s])`, lookback)
	rows, err := vmQuery(cfg, restartQ)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		if s.value <= 1 {
			continue
		}
		host := metricHost(s.metric)
		name := s.metric["name"]
		if name == "" {
			name = "unknown"
		}
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("container_restarts.%s.%s", host, name), Collector: "containers",
			Subject: host, Kind: "container_restarts", Value: int64(s.value), Unit: "restarts",
			Message:  fmt.Sprintf("%s: container %s restarted %d times in %s", host, name, int64(s.value), lookback),
			Evidence: map[string]any{"promql": restartQ, "grafana_url": cfg.GrafanaURL}, Severity: "info",
		})
	}

	oomQ := fmt.Sprintf("increase(container_oom_events_total[%s])", lookback)
	rows, err = vmQuery(cfg, oomQ)
	if err != nil {
		panic(err)
	}
	for _, s := range rows {
		if s.value < 1 {
			continue
		}
		host := metricHost(s.metric)
		name := s.metric["name"]
		if name == "" {
			name = "unknown"
		}
		result.Observations = append(result.Observations, &hr.Observation{
			ID: fmt.Sprintf("container_oom.%s.%s", host, name), Collector: "containers",
			Subject: host, Kind: "container_oom", Value: int64(s.value), Unit: "events",
			Message:  fmt.Sprintf("%s: container %s was OOM-killed %d times in %s", host, name, int64(s.value), lookback),
			Evidence: map[string]any{"promql": oomQ, "grafana_url": cfg.GrafanaURL}, Severity: "info",
		})
	}

	missingNames := make([]string, 0, len(missing))
	for _, k := range missing {
		missingNames = append(missingNames, k.host+"/"+k.name)
	}
	addedNames := make([]string, 0, len(added))
	for _, k := range added {
		addedNames = append(addedNames, k.host+"/"+k.name)
	}
	result.Data = &hr.ContainersData{
		CurrentCount: len(current), PreviousCount: len(previous),
		Missing: missingNames, New: addedNames,
	}
	return result
}

func keyLess(a, b struct{ host, name string }) bool {
	if a.host != b.host {
		return a.host < b.host
	}
	return a.name < b.name
}
